### PR TITLE
[SPARK-30857][SQL][2.4] Fix truncations of timestamps before the epoch to hours and days

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to replace `%` by `Math.floorMod` in `DateTimeUtils.truncTimestamp` for the `HOUR` and `DAY` levels.

### Why are the changes needed?
This fixes the issue of incorrect truncation of timestamps before the epoch `1970-01-01T00:00:00.000000Z` to the `HOUR` and `DAY` levels. For example, timestamps after the epoch are truncated by cutting off the rest part of the timestamp:
```sql
spark-sql> select date_trunc('HOUR', '2020-02-11 00:01:02.123'), date_trunc('HOUR', '2020-02-11 00:01:02.789');
2020-02-11 00:00:00	2020-02-11 00:00:00
```
but hours in the truncated timestamp before the epoch are increased by 1:
```sql
spark-sql> select date_trunc('HOUR', '1960-02-11 00:01:02.123'), date_trunc('HOUR', '1960-02-11 00:01:02.789');
1960-02-11 01:00:00	1960-02-11 01:00:00
```

### Does this PR introduce any user-facing change?
Yes. After the changes, the example above outputs correct result:
```sql
spark-sql> select date_trunc('HOUR', '1960-02-11 00:01:02.123');
1960-02-11 00:00:00
```

### How was this patch tested?
Added new tests to `DateFunctionsSuite`.

### CODE REVIEW

**Comment by MaxGekk** [comment]
@cloud-fan @dongjoon-hyun 2.4 has similar to #27611 issue but for the `HOUR` and `DAY` levels. Spark 3.0 doesn't have the problem because it uses Java 8 time API in truncations to hours and days.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
@MaxGekk can you fix the conflicts? Thanks!

File: sql/core/src/test/scala/org/apache/spark/sql/DateFunctionsSuite.scala [hunk 1]
```diff
@@ -729,4 +729,17 @@ class DateFunctionsSuite extends QueryTest with SharedSQLContext {
...
+    def checkTrunc(level: String, expected: String): Unit = {
+      val df = Seq("1961-04-12 00:01:02.345")
+        .toDF()
+        .select($"value".cast("timestamp").as("ts"))
+        .select(date_trunc(level, $"ts").cast("string"))
+      checkAnswer(df, Row(expected))
+    }
+
+    checkTrunc("HOUR", "1961-04-12 00:00:00")
+    checkTrunc("DAY", "1961-04-12 00:00:00")
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/DateFunctionsSuite.scala]: nit. If possible, can we have additional test statement for `minute` and `second` together to make it sure?

**Comment by MaxGekk** [comment]
jenkins, retest this, please

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
thanks, merging to 2.4!

**Comment by dongjoon-hyun** [comment]
Thank you, @MaxGekk and @cloud-fan .

**Pull request closed without merging**